feat(VCommandPalette): new components

MatthewAry opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://vuetifyjs.com/getting-started/contributing

Provide a general summary of your changes in the title above
Keep the title short and descriptive, as it will be used as a commit message
PR titles should follow conventional-changelog-angular:
https://vuetifyjs.com/getting-started/contributing/#commit-guidelines
-->

## Description
Adds a new VCommandPalette component

## Markup:
Use the `packages/vuetify/playgrounds/Playground.commandpalette.vue` file to play around with it.

<!-- Paste your FULL packages/vuetify/dev/Playground.vue here --->
```vue
<script setup lang="ts">
  import { ref } from 'vue'

  const model = ref(true)
  const showSnack = ref(false)
  const lastExecutedCommand = ref('')

  const items = [
    {
      title: 'New',
      value: 'new',
      prependIcon: 'mdi-plus',
      children: [
        {
          title: 'Project',
          prependIcon: 'mdi-folder-plus-outline',
          subtitle: 'Create a new project',
          value: 'new:project',
          hotkey: 'ctrl+alt+n',
        },
        {
          title: 'File',
          prependIcon: 'mdi-file-plus-outline',
          subtitle: 'Create a new file',
          value: 'new:file',
        },
        {
          title: 'User',
          prependIcon: 'mdi-account-plus-outline',
          subtitle: 'Create a new user',
          value: 'new:user',
        },
      ],
    },
    {
      title: 'Git',
      value: 'git',
      prependIcon: 'mdi-git',
      children: [
        {
          title: 'Commit',
          prependIcon: 'mdi-source-commit',
          subtitle: 'Commit changes',
          value: 'git:commit',
          hotkey: 'ctrl+k',
        },
        {
          title: 'Push',
          prependIcon: 'mdi-source-pull',
          subtitle: 'Push changes',
          value: 'git:push',
          hotkey: 'ctrl+shift+k',
        },
        {
          title: 'Fetch',
          prependIcon: 'mdi-source-branch',
          subtitle: 'Fetch changes',
          value: 'git:fetch',
        },
      ],
    },
    {
      title: 'File: Save',
      prependIcon: 'mdi-content-save',
      subtitle: 'Save the current file',
      value: 'file:save',
      hotkey: 'ctrl+s',
    },
    {
      title: 'File: Save As...',
      prependIcon: 'mdi-content-save-outline',
      subtitle: 'Save the current file as...',
      value: 'file:saveAs',
      hotkey: 'ctrl+shift+s',
    },
    {
      title: 'Preferences',
      value: 'preferences',
      prependIcon: 'mdi-cog',
      children: [
        {
          title: 'Settings',
          prependIcon: 'mdi-cog-outline',
          subtitle: 'Open settings',
          value: 'pref:settings',
        },
        {
          title: 'Keyboard Shortcuts',
          prependIcon: 'mdi-keyboard-settings-outline',
          subtitle: 'Configure keyboard shortcuts',
          value: 'pref:keys',
        },
        {
          title: 'Theme',
          prependIcon: 'mdi-palette-outline',
          value: 'pref:theme',
          children: [
            {
              title: 'Dark Theme',
              prependIcon: 'mdi-weather-night',
              value: 'theme:dark',
            },
            {
              title: 'Light Theme',
              prependIcon: 'mdi-weather-sunny',
              value: 'theme:light',
            },
          ],
        },
      ],
    },
    {
      title: 'Find',
      prependIcon: 'mdi-magnify',
      subtitle: 'Find in the current file',
      value: 'find',
      hotkey: 'ctrl-f',
    },
    {
      title: 'Find in Files',
      prependIcon: 'mdi-file-find-outline',
      subtitle: 'Find in the entire workspace',
      value: 'find:files',
    },
  ]

  async function onItemClick (item: any) {
    lastExecutedCommand.value = item.title
    showSnack.value = true
  }
</script>

<template>
  <v-btn text="Open" @click="model = !model" />
  <v-command-palette
    v-model="model"
    :item-props="true"
    :items="items"
    hotkey="cmd+k"
    @click:item="onItemClick"
  />
  <v-snackbar
    v-model="showSnack"
    :timeout="2000"
    color="blue-grey"
    rounded="pill"
  >
    Executed command: <strong>{{ lastExecutedCommand }}</strong>.
  </v-snackbar>
</template>
```

MatthewAry committed:
No warnings. Some progress. Time for the real work.

MatthewAry committed:
feat(VCommandPalette): add structured items and hotkey support

- Added a comprehensive list of command palette items with titles, subtitles, and icons.
- Implemented hotkey functionality for various commands, improving user accessibility.
- Enhanced the VCommandPalette component to support item click events and dynamic rendering of items.
- Updated styles for better visual presentation of the command palette list.
- Introduced localization for command palette instructions in English.

MatthewAry committed:
feat(VCommandPalette): better hotkey handling

- Improved VHotkey component for better hotkey representation.
- Updated useHotkey function to include preventDefault option for more control over event handling.
- Enhanced VCommandPalette to support new slot types for header and footer, improving customization.
- Added autofocus to the VCommandPaletteSearch component for better user experience.
- Refactored event handling in VCommandPalette to utilize updated hotkey options.

MatthewAry committed:
feat(VCommandPalette): enhance item structure and type safety

- Introduced a comprehensive type system for command palette items, including action, link, parent, and group definitions.
- Updated the VCommandPaletteList component to utilize the new item types, improving type safety and clarity.
- Refactored item handling logic to accommodate the new structure, ensuring compatibility with existing functionality.
- Enhanced the example usage to demonstrate the new item definitions and their properties.
- Improved prop types for better integration with VList, ensuring a more robust and maintainable codebase.

MatthewAry committed:
Comments

MatthewAry committed:
feat(VCommandPalette): improve hotkey and fix item rendering

- Updated VCommandPalette to improve hotkey representation by changing the format from 'ctrl+f' to 'ctrl-f'.
- Simplified item rendering by removing unnecessary template slots for VListItem, enhancing code readability.
- Enhanced VHotkey component with a more structured key mapping and improved display logic for key combinations.
- Added localization support for the new 'then' hotkey term in English.
- Refactored styles in VHotkey for better visual consistency and accessibility.

MatthewAry committed:
Apply feedback from @johnleider

MatthewAry committed:
Invoke handler, tweak to VHotkey

MatthewAry committed:
feat(VCommandPalette): item groups and better navigation

- Improved the command palette by adding structured item types, including groups and dividers, for better organization and accessibility.
- Updated the handler functions to use TypeScript for type safety and clarity.
- Enhanced the VHotkey component with a more comprehensive key mapping and improved display logic for key combinations.
- Refactored item rendering logic in VCommandPaletteList to support new item structures, ensuring a more maintainable and readable codebase.
- Added support for selectable items count to improve keyboard navigation experience.
- Improved styling in VCommandPalette for better visual consistency and user experience.

MatthewAry committed:
Toggle Theme

MatthewAry committed:
Better search

MatthewAry committed:
Pesky transition

MatthewAry committed:
A variety of improvement

MatthewAry committed:
Formatting

MatthewAry committed:
Improved VHotkey and Added JSDoc to mergeDeep

User is able to selectively override items in the keymap in VHotkey

MatthewAry committed:
feat(hotkey): add tests for useHotkey composable

- Introduced a new `useHotkey` composable to handle keyboard shortcuts and key sequences in Vue applications.
- Supports single key combinations and complex key sequences with customizable options for event type, input handling, and default behavior prevention.
- Implemented platform-specific key handling for Mac users, ensuring consistent behavior across different operating systems.
- Added comprehensive unit tests to validate functionality, including edge cases for key sequences and modifier keys.
- Enhanced accessibility by allowing hotkeys to be ignored when focused on input elements, configurable via options.

MatthewAry committed:
feat(VCommandPalette): enhance filtering logic and slot scopes

- Introduced a custom filter function for command palette items that accommodates group and parent hierarchies, improving search accuracy.
- Updated the `prepend` and `append` slot scopes to include a `search` property, enhancing accessibility and usability.
- Refactored filtering logic to maintain compatibility with the existing `useFilter` composable while implementing specific group filtering behavior.
- Improved code readability and maintainability by leveraging TypeScript for type definitions and ensuring proper naming conventions.

MatthewAry committed:
Update playground

MatthewAry committed:
feat(VCommandPalette): refine slot naming and enhance selection logic

- Updated slot names from `prepend-item` and `append-item` to `prepend-list` and `append-list` for better clarity and consistency.
- Enhanced the selection logic to auto-select the first item when the filter changes or when the dialog opens, improving user experience.
- Adjusted SCSS styles for the list divider to include a margin-bottom for better spacing.
- Ensured code readability and maintainability by adhering to TypeScript conventions and proper naming practices.

MatthewAry committed:
feat(VCommandPalette): make selections feel snappy

- Added transition removal for the list item overlay to improve performance and visual consistency.
- Updated selection logic for arrow key navigation to clarify behavior with comments, ensuring better maintainability.
- Changed scroll behavior from 'smooth' to 'instant' for immediate feedback during selection.
- Deleted the unused VCommandPaletteInstructions.scss file

MatthewAry committed:
Some tests aren't passing. They're tricky!

MatthewAry committed:
Allow VDialog to unload core VCommandPalette logic. All tests pass.

MatthewAry committed:
feat(VCommandPalette): enhance accessibility and navigation features

- Updated the VCommandPalette to include hotkeys for dark and light theme toggling, improving user experience and accessibility.
- Refactored the useHotkey composable to handle edge cases in SSR environments, ensuring robust functionality across different setups.
- Enhanced the VCommandPalette's item registration and navigation logic, allowing for better keyboard navigation and focus management.
- Improved ARIA attributes for better screen reader support, ensuring compliance with accessibility standards.
- Added comprehensive tests to validate the new features and ensure consistent behavior.

MatthewAry committed:
feat(VCommandPalette): enhance item matching and state management

- Updated the item matching logic to include keywords, improving search accuracy for command palette items.
- Modified filter keys to encompass 'title', 'subtitle', and 'keywords' for more comprehensive filtering.
- Implemented state reset functionality when the parent dialog closes, ensuring a clean slate for subsequent openings.
- Enhanced code readability by using TypeScript for type definitions and ensuring proper naming conventions.
- Added necessary imports and adjusted existing logic for better maintainability and accessibility compliance.

MatthewAry committed:
Comments

MatthewAry committed:
feat(VCommandPalette): improve item typing and transformation logic

- Added proper typing for the `items` prop in `VCommandPalette`, ensuring it accepts an array of `VCommandPaletteItem`.
- Refactored item processing logic to directly utilize `VCommandPaletteItem` properties, enhancing clarity and maintainability.
- Updated import statements for consistency and clarity, renaming `VCommandPaletteItem` to `VCommandPaletteItemComponent` for better distinction.
- Enhanced the `VListDisplayTypes` type definition for improved type safety and readability.
- Ensured accessibility compliance by maintaining focus restoration and hotkey registration logic.

MatthewAry committed:
Localization and custom-layout example

MatthewAry committed:
Remove unused imports

MatthewAry committed:
feat(VCommandPalette): enhance VHotkey functionality and testing

- Introduced improved parsing logic for key sequences, specifically handling the minus key as a separator only between alphanumeric characters.
- Added alternative names for the minus key, allowing for better user experience and flexibility in key mappings.
- Implemented comprehensive unit tests for VHotkey, covering various key combinations, display modes, and edge cases to ensure robust functionality.
- Refactored VCommandPaletteList to adjust the placement of the end divider for improved visual consistency.
- Updated scroll behavior for keyboard navigation to 'auto' for a smoother user experience.
- Ensured code quality and maintainability by adhering to TypeScript standards and enhancing accessibility features.

MatthewAry committed:
refactor(VCommandPalette): no const functions

- Changed the `transformFilteredItems` and `processItems` functions from arrow functions to named functions

MatthewAry committed:
refactor(VCommandPalette): replace ref with shallowRef

- Updated the `navigationMode` and `selectedIndex` properties to use `shallowRef` instead of `ref` for improved reactivity and performance.
- Removed unused `ref` import from `useCommandPaletteContext` and adjusted related code for consistency.

MatthewAry committed:
Simplify

MatthewAry committed:
Merge remote-tracking branch 'origin/dev' into branch

MatthewAry committed:
refactor(VHotkey): integrate VKbd component for key display

- Updated the VHotkey component to utilize the new VKbd component.
- Adjusted SCSS to apply specific styles to the VKbd component.

MatthewAry committed:
feat(iconsets): add keyboard-related icon aliases

- Introduced new icon aliases for keyboard keys (command, ctrl, shift, alt, enter, arrow keys, backspace) in md.ts, mdi-svg.ts, and mdi.ts files.
- Updated VHotkey component to utilize these new aliases.

MatthewAry committed:
fix Some weird character in keymap

MatthewAry committed:
Remove playground

MatthewAry committed:
Fix test

MatthewAry committed:
Localize VHotkey

MatthewAry committed:
Merge branch 'master' into matthewary/v-command-palette

MatthewAry committed:
Use VHotkey from labs

MatthewAry committed:
Feat: Add contained prop

- Use VCommandPalette without VDialog. Necessary for documentation purposes.
- Preliminary Plan for component documentation

MatthewAry committed:
Merge branch 'master' into matthewary/v-command-palette

J-Sek committed:
chore: export sub-components

J-Sek committed:
fix: custom layout example

J-Sek committed:
fix: move hooks outside of watch

J-Sek committed:
fix tests

J-Sek commented on packages/vuetify/src/locale/en.ts:
```diff
@@ -117,6 +117,13 @@ export default {
+    navigate: 'to navigate',
+    goBack: 'to go back',
+    close: 'to close',
+    placeholder: 'Type a command or search...',
```
missing "then"?

J-Sek committed:
lint: missing comma

J-Sek committed:
fix tests

J-Sek committed:
use `wait` in tests

J-Sek committed:
delay reset until dialog dialog disappears + expose

J-Sek committed:
refactor props selection

J-Sek committed:
fix missing locale entry

J-Sek commented:
Bug:
- when I move or hover an item and use keyboard Tab, there are 2 selected items that move up/down in parallel

Missing feature:
- when I click an item to navigate into a group, there is no element to click to go back. I am forced to use the keyboard "backspace". I know that we intend this whole component for power users glued to the keyboard, but it feels like a thing that would complete the picture (and also help indicate to the user that they navigated "inside" and can go back).
  - maybe a `$prev` icon could replace `$search` (that should be a `prepend-inner` icon of the search field by default)

J-Sek committed:
chore: easier customization with `list-props` and `search-props`

J-Sek committed:
chore: expose VSheet props instead of VDialog

J-Sek committed:
fix: adapt subheader color to the background

J-Sek committed:
fix: consistently apply class to VSheet

MatthewAry committed:
Fix nav

MatthewAry committed:
Reduce LOC. Fix tests.

MatthewAry committed:
Docs! First pass

MatthewAry committed:
Merge 'origin/master' into matthewary/v-command-palette

MatthewAry committed:
HOW DO YOU WRITE DOCS FOR THIS?

MatthewAry committed:
Fixes to docs

MatthewAry committed:
Merge remote-tracking branch 'origin/master'  into branch

MatthewAry committed:
Add hotkeys scope and snackbar functionality to command palette examples

- Introduced `hotkeys-scope="focused"` to various command palette examples for improved hotkey management.
- Implemented snackbar notifications in the hierarchical example to provide user feedback on item actions.
- Updated the VCommandPalette component to support scoped hotkeys and enhanced item click handling for better user experience.
- Added left/right navigation support for grid layouts in the command palette navigation logic.

MatthewAry committed:
Add back navigation and subheader support to VCommandPalette

- Introduced `showSearchBackButton` prop to control the visibility of the back button in the search input during nested navigation.
- Enhanced `VCommandPaletteSearch` to display a back button instead of the search icon when navigating into parent items.
- Added a subheader slot to `VCommandPalette` that provides contextual information and allows back navigation.
- Implemented tests for back navigation and subheader functionality to ensure proper behavior in various scenarios.
- Updated navigation logic to support keyboard navigation with Tab and Shift+Tab for accessibility.

MatthewAry committed:
Some improvements

MatthewAry committed:
Merge branch 'master' into matthewary/v-command-palette

MatthewAry commented:
I will do another pass at some point and clean up comments.

J-Sek committed:
chore: improved usage example

J-Sek committed:
chore: simplified example with item slot

J-Sek committed:
fix: minor correction to font size for hotkeys

J-Sek committed:
chore: drop dialog props that does not seem to make any difference

J-Sek committed:
refactor: simplify main content slots

J-Sek committed:
chore: refactor search field

johnleider commented on packages/vuetify/src/composables/hotkey.ts:
```diff
@@ -0,0 +1,176 @@
+// Utilities
```
If this is a standalone it should be in its own PR

J-Sek replied:
Just a duplicate to remove.

johnleider commented on packages/vuetify/src/labs/VCommandPalette/VCommandPaletteInstructions.tsx:
```diff
@@ -0,0 +1,105 @@
+ */
+export const makeVCommandPaletteInstructionsProps = propsFactory({
+  // Whether there are items to select (shows Enter and arrow key instructions)
+  hasItems: Boolean,
```
What do we call this for things like list-items and treeview? We have no props in the framework that would align with this naming convention.

johnleider commented on packages/vuetify/src/labs/VCommandPalette/VCommandPaletteInstructions.tsx:
```diff
@@ -0,0 +1,105 @@
+  hasParent: Boolean,
+  // ID for ARIA describedby relationships
+  id: String,
+  ...makeThemeProps(),
```
What do we call this for things like list-items and treeview? We have no props in the framework that would align with this naming convention.

johnleider commented on packages/vuetify/src/labs/VCommandPalette/VCommandPaletteItem.tsx:
```diff
@@ -0,0 +1,155 @@
+  // The item data to render
+  item: {
+    type: Object as PropType<any>,
+    required: true,
```
No required props.

johnleider commented on packages/vuetify/src/labs/VCommandPalette/VCommandPaletteItem.tsx:
```diff
@@ -0,0 +1,155 @@
+  // The item's index in the selectable items list
+  index: {
+    type: Number,
+    required: true,
```
No required props.

johnleider commented on packages/vuetify/src/labs/VCommandPalette/VCommandPaletteItem.tsx:
```diff
@@ -0,0 +1,155 @@
+    required: true,
+  },
+  // The HTML tag to render as the root element
+  tag: {
```
This is handled by makeComponentProps, just do `makeComponentProps({ tag: 'div' })`

johnleider commented on packages/vuetify/src/labs/VCommandPalette/VCommandPaletteItem.tsx:
```diff
@@ -0,0 +1,155 @@
+  setup (props, { slots }) {
+    // Get the command palette context for registration and state management
+    const context = useCommandPaletteContext()
+    const elementRef = ref<HTMLElement>()
```
Should we be using `useTemplateRef`?

johnleider commented on packages/vuetify/src/labs/VCommandPalette/VCommandPaletteItems.tsx:
```diff
@@ -0,0 +1,89 @@
+  // The HTML tag to render as the root element
+  tag: {
+    type: String,
+    default: 'div',
```
This is handled by makeComponentProps, just do `makeComponentProps({ tag: 'div' })`

johnleider reviewed (changes requested):
Lots of comments need to go the way of the samurai. I didn't really get through everything. This P.R. is simply massive. Is there a way that we can cut off some pieces and get them into the framework separately so that we can reduce it's scope of responsibility?

Overall, code looks great, awesome job 👍

J-Sek commented on packages/docs/src/examples/v-command-palette/contained.vue:
```diff
@@ -0,0 +1,87 @@
+      :items="items"
+      :model-value="true"
+      class="align-self-stretch mx-auto"
+      hotkeys-scope="focused"
```
I have opened the docs page and `contained` examples still hijack keyboard up/down/backspace.

J-Sek committed:
chore: drop default clearable

J-Sek committed:
feat: add `item.append` and `search.append` slots

J-Sek committed:
feat: accept `item-title` + slots for title, prepend & subheader

J-Sek committed:
fix main hotkey (drop duplicate implementation)

J-Sek committed:
fix type import
